aten/src/ATen/native/MaxUnpooling.cpp

```diff
@@ -0,0 +1,602 @@
+            oheight);
+      }));
+  return output;
+};
+
+Tensor max_unpooling2d_forward_cpu(
+    const Tensor& self,
+    const Tensor& indices,
+    IntList output_size) {
+  auto output = at::empty({0}, self.options());
+  max_unpooling2d_forward_out_cpu(output, self, indices, output_size);
+  return output;
+}
+
+template <typename scalar_t>
+Tensor max_unpooling3d_forward_out_cpu_frame(
+    Tensor& output,
+    const Tensor& input,
+    const Tensor& indices,
+    int64_t oT,
+    int64_t oW,
+    int64_t oH,
+    int64_t dT,
+    int64_t dW,
+    int64_t dH,
+    int64_t pT,
+    int64_t pW,
+    int64_t pH) {
+  int64_t nBatch = 1;
+  int64_t dimw = 3;
+  int64_t dimh = 2;
+  int64_t dimt = 1;
+
+  if (input.ndimension() == 5) {
+    nBatch = input.size(0);
+    dimw++;
+    dimh++;
+    dimt++;
+  }
+
+  int64_t nSlices = input.size(dimt - 1);
+  int64_t iT = input.size(dimt);
+  int64_t iH = input.size(dimh);
+  int64_t iW = input.size(dimw);
+
+  scalar_t* input_data = input.data<scalar_t>();
+  scalar_t* output_data = output.data<scalar_t>();
+  int64_t* indices_data = indices.data<int64_t>();
+
+  for (int64_t p = 0; p < nBatch; p++) {
+    int64_t inputOffset = p * nSlices * iT * iW * iH;
+    int64_t outputOffset = p * nSlices * oT * oW * oH;
+    int64_t k;
+    bool has_error = false;
+    int error_index = 0;
+#pragma omp parallel for private(k)
+    for (k = 0; k < nSlices; k++) {
+      int64_t finalInputOffset = inputOffset + k * iT * iW * iH;
+      int64_t finalOutputOffset = outputOffset + k * oT * oW * oH;
+
+      scalar_t* output_p_k = output_data + finalOutputOffset;
+      scalar_t* input_p_k = input_data + finalInputOffset;
+      int64_t* ind_p_k = indices_data + finalInputOffset;
+      int maxp;
+      for (int64_t t = 0; t < iT; t++) {
+        for (int64_t i = 0; i < iH; i++) {
+          for (int64_t j = 0; j < iW; j++) {
+            int64_t index = t * iH * iW + i * iW + j;
+            maxp = ind_p_k[index];
+            if (maxp < 0 || maxp >= oT * oW * oH) {
+#pragma omp critical
+              {
+                has_error = true;
+                error_index = maxp;
+              }
+            } else {
+              output_p_k[maxp] = input_p_k[index];
+            }
+          }
+        }
+      }
+      if (has_error) {
+        AT_ERROR(
+            "found an invalid max index ",
+            error_index,
+            " (output volumes are of size ",
+            oT,
+            "x",
+            oH,
+            "x",
+            oW);
+      }
+    }
+  }
+  return output;
+}
+
+void max_unpooling3d_shape_check(
+    const Tensor& input,
+    const Tensor& gradOutput,
+    const Tensor& indices,
+    IntList output_size,
+    IntList stride,
+    IntList padding) {
+  int64_t oT = output_size[0];
+  int64_t oW = output_size[1];
+  int64_t oH = output_size[2];
+  AT_CHECK(
+      indices.scalar_type() == at::ScalarType::Long,
+      "elements in indices should be type Long");
```

Comment:
nit: We really should use int64 here, it's the modern designation that we're encouraging people to use compared to long (e.g., torch.zeros(2, dtype=torch.int64))